yetanotherco/aligned_layer
# fix: batcher sends proof even if eip712 signature contents are incompatible

# This PR
There was a bug that the batcher accepted a proof if incompatible `--chain` and `--batcher-payments-address` where provided, which resulted in an Ethereum tx revert for the whole batch submition.

Resolving this, it was found that these 2 parameters where redundant, so they were both removed and replaced for the `--network` , which accepts either `devnet`, `holesky-stage` or `holesky` .

Then, this parameter is used by the SDK to the appropriate contract addresses, while the chainId is resolved from the RPC (this has not changed, which also lead to confusion).

## To Deploy
- As the SDK changes with this PR, a new release with a new tag should be made (maybe v0.7.1 or v0.8.0)
- The Batcher should be redeployed, since a new check is being made
  - `if client_msg.verification_data.payment_service_addr != self.payment_service.address() {`
- The `task sender` should be restarted using this new parameter (make targets have been migrated)
- ZKQuiz contract has also changed so it must be redeployed and its address repointed to the new on-chain one

## To Test
You can test the following work correctly:
- the new task senders
- the batcher
- the examples
  - Validating Public Inputs can be tested on testnet, as normal (`make submit_fibonacci_proof`)
  - ZKQuiz changed a value in the contract so it should be tested on devnet before deploying it on Staging

### Test ZKQuiz in devnet
1) Run anvil
```
make anvil_start_with_block_time
```

2) Deploy ZKQuiz verifier contract:
```
cd examples/zkquiz
make deploy_verifier
```

3) Configure variables in `examples/zkquiz/contracts/.env`:
```
RPC_URL=http://localhost:8545
PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
ALIGNED_SERVICE_MANAGER_ADDRESS=0x1613beB3B2C4f22Ee086B2b38C1476A3cE7f78E8
BATCHER_PAYMENT_SERVICE_ADDRESS=0x7969c5eD335650692Bc04293B07F5BF2e7A673C0
```

4) And configure variables in `examples/zkquiz/Makefile/.env`: (double check `CONTRACT_ADDRESS` is the same address as outputed when deployed the zkquiz verifier contract)
```
CONTRACT_ADDRESS=0xc351628EB244ec633d5f21fBD6621e1a683B1181
RPC_URL=http://localhost:8545
KEYSTORE_PATH=../../../../config-files/devnet/keys/anvil.ecdsa.key.json
```

5) Change some code in `examples/zkquiz/quiz/script/src/main.rs`"
- line 20, to:
```
const BATCHER_URL: &str = "ws://localhost:8080";
```
- line 23, to:
```
const NETWORK: Network = Network::Devnet;
```
- line 33, to:
```
        default_value = "http://localhost:8545"
```
- line 49, to:
```
.with_chain_id(31337u64);
```

6) Run zkquiz:
```
make answer_quiz
```

Solves https://github.com/yetanotherco/aligned_layer/issues/999
